1-Parallelize-your-python-code_Futures_and_Delayed.ipynb

```diff
@@ -0,0 +1,1062 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f772bf99-a14a-40c0-b7bd-1da131f6dff0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "res[:3]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6d37347b-8f8b-459e-be1a-e3a0da5169e0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "r = dask.compute(res)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6b91a94b-40f1-4312-88ea-1506b405dee8",
+   "metadata": {},
+   "source": [
+    "## Waiting on Futures: `wait` and `as_completed` \n",
+    "\n",
+    "You can wait on a future or collection of futures using the `wait` function, which blocks until all futures are finished or have erred. This is useful when you need the all the futures to be completed to proceed with your computations. \n",
+    "\n",
+    "In other situations you might need to iterate over the futures as they complete, to do so you will use the `as_completed` function.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a50089dc-20ca-44d7-9c45-c3fd2a7814d3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from dask.distributed import as_completed, wait"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3bd0d65c-eac1-45c2-b487-8574cc34c777",
+   "metadata": {},
+   "source": [
+    "### `wait()`\n",
+    "\n",
+    "Notice that the cell below will not block until the futures end, it will trigger the computation and let you proceed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f4878a23-143d-4c30-aadd-6a535c0bcf97",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fut_example = client.submit(scrape_tag, 100)\n",
+    "fut_example"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6715063a-a499-4496-ba3d-dd22e16a0042",
+   "metadata": {},
+   "source": [
+    "If you want to wait:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f7d5c19-5fde-4474-866c-b7d1b3711592",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "some_futures = client.map(scrape_tag, range(50, 90))\n",
+    "wait(some_futures)\n",
+    "some_futures[-1]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7717095f-5169-4ba3-9dc4-8acd25e2ebaf",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#clear all the futures that are around\n",
+    "client.restart()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e1d08e82-47eb-4b55-9a40-a5cd11dddf5b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "futures = client.map(scrape_tag, range(1, 33))\n",
+    "\n",
+    "tot_views= 0 \n",
+    "\n",
+    "for future in as_completed(futures):\n",
+    "    views = future.result()['views']\n",
+    "    tot_views += views.sum()\n",
+    "\n",
+    "tot_views"
+   ]
+  },
```

Comment:
This doesn't seem like a good use case for `as_completed`.  There's no reason not to just use `client.gather` here.  I think it's important that we not just show the use of an API, but also have them experience why that system is important and what it can do for them.  They should walk away thinking "Ah!  That's just like this problem I have at work!  I can use this tool there".  This doesn't give them that example I don't think.

Reply:
A suggestion!  

Scraping is often accompanied by *crawling*.   You scrape a page, find all of the links, scrape those pages, find all of the links in those, scrape further, etc..  We could do the same.  

```python
futures = [client.submit(scrape_and_crawl, url)]
crawling = as_completed(futures)
seen = set()
for future in crawling:
    result, links = future.result()
    for link in links:
        if link not in seen:
            future = client.submit(scrape_and_crawl, link)
            crawling.add(link)
            seen.add(link)
    # do something with result
```

This shows off dynamism, which opens lots of doors.  It also is more interesting I think.